base/genericmemory.jl

```diff
@@ -288,3 +288,19 @@ function indcopy(sz::Dims, I::GenericMemory)
     src = eltype(I)[I[i][_findin(I[i], i < n ? (1:sz[i]) : (1:s))] for i = 1:n]
     dst, src
 end
+
+# Wrapping a memory region in an Array
+@eval function reshape(m::GenericMemory{M, T}, dims::Vararg{Int, N}) where {M, T, N}
+    len = Core.checked_dims(dims...)
+    length(m) == len || throw(DimensionMismatch("parent has $(length(m)) elements, which is incompatible with size $(dims)"))
+    ref = MemoryRef(m)
+    $(Expr(:new, :(Array{T, N}), :ref, :dims))
+end
+
+@eval function view(m::GenericMemory{M, T}, inds::Union{UnitRange, OneTo}) where {M, T}
+    !(inds isa OneTo) && isempty(inds) && return T[] # needed to allow view(Memory{T}(undef, 0), 2:1)
```

Comment:
Why doesn't it matter that OneTo can be empty?
```suggestion
    isempty(inds) && return T[] # needed to allow view(Memory{T}(undef, 0), 2:1)
```

You can always construct `MemoryRef(m)`, but can't always then offset it by `MemoryRef(ref, first(inds))`

Reply:
I was thinking of the following implementation which I believe is safe and should be faster:

```julia
@eval function view(m::GenericMemory{M, T}, inds::UnitRange) where {M, T}
    isempty(inds) && return T[] # needed to allow view(Memory{T}(undef, 0), 2:1)
    @boundscheck checkbounds(m, inds)
    ref = @inbounds MemoryRef(m, first(inds)) # @inbounds here is safe because of the checkbounds and isempty checks above
    dims = (length(inds),)
    $(Expr(:new, :(Array{T, 1}), :ref, :dims))
end
@eval function view(m::GenericMemory{M, T}, inds::OneTo) where {M, T}
    @boundscheck checkbounds(m, inds)
    ref = MemoryRef(m) # Does not perform bounds checking, but empty memory gets special handling to make this okay
    dims = (length(inds),)
    $(Expr(:new, :(Array{T, 1}), :ref, :dims))
end
```

But that implementation doesn't benchmark as faster for me so I'll just go with the simpler one in this PR with more checks.